From 7ed041fbdb1b0ccf3bc7efac8bfbd314ef3db1c1 Mon Sep 17 00:00:00 2001
From: Phantom X <PhantomX@users.noreply.github.com>
Date: Fri, 3 Sep 2021 22:01:03 -0300
Subject: [PATCH] Change savefiles location

Ugly dirty fix to save to configdir/sav|mch directory instead rom file location
---
 src/frontend/Util_ROM.cpp        | 34 +++++++++++++++++++++++++++++---
 src/frontend/qt_sdl/Platform.cpp |  4 ++++
 src/frontend/qt_sdl/main.cpp     |  9 +++++++--
 3 files changed, 42 insertions(+), 5 deletions(-)

diff --git a/src/frontend/Util_ROM.cpp b/src/frontend/Util_ROM.cpp
index 19023fb..1f8a7fd 100644
--- a/src/frontend/Util_ROM.cpp
+++ b/src/frontend/Util_ROM.cpp
@@ -21,6 +21,10 @@
 
 #include <utility>
 
+#include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+
 #ifdef ARCHIVE_SUPPORT_ENABLED
 #include "ArchiveUtil.h"
 #endif
@@ -232,8 +236,29 @@ int SetupDSiNAND()
     return Load_OK;
 }
 
+// https://thispointer.com/c-how-to-get-filename-from-a-path-with-or-without-extension-boost-c17-filesytem-library/
+std::string getFileName(std::string filePath, char separator = '/')
+{
+    // Get last dot position
+    std::size_t dotPos = filePath.rfind('.');
+    std::size_t sepPos = filePath.rfind(separator);
+
+    if(sepPos != std::string::npos)
+    {
+        return filePath.substr(sepPos + 1, filePath.size());
+    }
+    return "";
+}
+
 void LoadCheats()
 {
+    std::string homedir = getenv("HOME");
+    if (homedir.empty())
+    {
+         std::string homedir = getpwuid(getuid())->pw_dir;
+    }
+    std::string mchpath = homedir + "/.config/melonDS/mch/";
+
     if (CheatFile)
     {
         delete CheatFile;
@@ -243,13 +268,16 @@ void LoadCheats()
     char filename[1024];
     if (ROMPath[ROMSlot_NDS][0] != '\0')
     {
-        strncpy(filename, ROMPath[ROMSlot_NDS], 1023);
+        std::string fullname = mchpath + getFileName(ROMPath[ROMSlot_NDS]);
+        strncpy(filename, fullname.c_str(), 1023);
         filename[1023] = '\0';
-        strncpy(filename + strlen(ROMPath[ROMSlot_NDS]) - 3, "mch", 3);
+        strncpy(filename + strlen(fullname.c_str()) - 3, "mch", 3);
     }
     else
     {
-        strncpy(filename, "firmware.mch", 1023);
+        std::string fullname = mchpath + "firmware.mch";
+        filename[1023] = '\0';
+        strncpy(filename, fullname.c_str(), 1023);
     }
 
     // TODO: check for error (malformed cheat file, ...)
diff --git a/src/frontend/qt_sdl/Platform.cpp b/src/frontend/qt_sdl/Platform.cpp
index 81d86d3..f2ff2b3 100644
--- a/src/frontend/qt_sdl/Platform.cpp
+++ b/src/frontend/qt_sdl/Platform.cpp
@@ -108,6 +108,8 @@ void Init(int argc, char** argv)
     QString confdir;
     QDir config(QStandardPaths::writableLocation(QStandardPaths::ConfigLocation));
     config.mkdir("melonDS");
+    config.mkdir("melonDS/mch");
+    config.mkdir("melonDS/sav");
     confdir = config.absolutePath() + "/melonDS/";
     EmuDirectory = new char[confdir.length() + 1];
     memcpy(EmuDirectory, confdir.toUtf8().data(), confdir.length());
@@ -179,6 +181,8 @@ FILE* OpenLocalFile(const char* path, const char* mode)
         // Check user configuration directory
         QDir config(QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation));
         config.mkdir("melonDS");
+        config.mkdir("melonDS/mch");
+        config.mkdir("melonDS/sav");
         fullpath = config.absolutePath() + "/melonDS/";
         fullpath.append(path);
 #endif
diff --git a/src/frontend/qt_sdl/main.cpp b/src/frontend/qt_sdl/main.cpp
index 73fcf5a..102e7d3 100644
--- a/src/frontend/qt_sdl/main.cpp
+++ b/src/frontend/qt_sdl/main.cpp
@@ -41,6 +41,8 @@
 #include <sys/socket.h>
 #include <signal.h>
 #endif
+#include <QStandardPaths>
+#include <QDir>
 
 #include <SDL2/SDL.h>
 
@@ -1785,7 +1787,8 @@ void MainWindow::dropEvent(QDropEvent* event)
         else
         {
             slot = (romFileName.endsWith(".gba", Qt::CaseInsensitive) ? 1 : 0);
-            QString sramFileName = QFileInfo(_filename).absolutePath() + QDir::separator() + QFileInfo(romFileName).completeBaseName() + ".sav";
+            QDir config(QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation));
+            QString sramFileName = config.absolutePath() + "/melonDS/sav/" + QFileInfo(romFileName).completeBaseName() + ".sav";
 
             if(slot == 0)
                 strncpy(Frontend::NDSROMExtension, QFileInfo(romFileName).suffix().toStdString().c_str(), 4);
@@ -1877,10 +1880,12 @@ void MainWindow::loadROM(QByteArray *romData, QString archiveFileName, QString r
     recentFileList.prepend(archiveFileName);
     updateRecentFilesMenu();
 
+    QDir config(QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation));
+
     // Strip entire archive name and get folder path
     strncpy(Config::LastROMFolder, QFileInfo(archiveFileName).absolutePath().toStdString().c_str(), 1024);
 
-    QString sramFileName = QFileInfo(archiveFileName).absolutePath() + QDir::separator() + QFileInfo(romFileName).completeBaseName() + ".sav";
+    QString sramFileName = config.absolutePath() + "/melonDS/sav/" + QFileInfo(romFileName).completeBaseName() + ".sav";
 
     int slot; int res;
     if (romFileName.endsWith("gba"))
-- 
2.31.1

